### PR TITLE
[SPARK-31381][SPARK-29245][SQL] Upgrade built-in Hive 2.3.6 to 2.3.7

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

**Hive 2.3.7** fixed these issues:
- HIVE-21508: ClassCastException when initializing HiveMetaStoreClient on JDK10 or newer
- HIVE-21980:Parsing time can be high in case of deeply nested subqueries
- HIVE-22249: Support Parquet through HCatalog

### Why are the changes needed?
Fix CCE during creating HiveMetaStoreClient in JDK11 environment: [SPARK-29245](https://issues.apache.org/jira/browse/SPARK-29245).

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?

- [x] Test Jenkins with Hadoop 2.7 (https://github.com/apache/spark/pull/28148#issuecomment-616757840)
- [x] Test Jenkins with Hadoop 3.2 on JDK11 (https://github.com/apache/spark/pull/28148#issuecomment-616294353)
- [x] Manual test with remote hive metastore.

Hive side:

```
export JAVA_HOME=/usr/lib/jdk1.8.0_221
export PATH=$JAVA_HOME/bin:$PATH
cd /usr/lib/hive-2.3.6 # Start Hive metastore with Hive 2.3.6
bin/schematool -dbType derby -initSchema --verbose
bin/hive --service metastore
```

Spark side:

```
export JAVA_HOME=/usr/lib/jdk-11.0.3
export PATH=$JAVA_HOME/bin:$PATH
build/sbt clean package -Phive -Phadoop-3.2 -Phive-thriftserver
export SPARK_PREPEND_CLASSES=true
bin/spark-sql --conf spark.hadoop.hive.metastore.uris=thrift://localhost:9083
```

### CODE REVIEW

**Comment by wangyum** [comment]
retest this please

File: pom.xml [hunk 1]
```diff
@@ -275,6 +275,17 @@
         <enabled>false</enabled>
       </snapshots>
     </repository>
+    <repository>
```
Comment (wangyum) [inline pom.xml:278]: Will revert this change after Hive 2.3.7 is released.

File: project/SparkBuild.scala [hunk 2]
```diff
@@ -230,6 +230,9 @@ object SparkBuild extends PomBuild {
       DefaultMavenRepository,
       Resolver.mavenLocal,
       Resolver.file("local", file(Path.userHome.absolutePath + "/.ivy2/local"))(Resolver.ivyStylePatterns)
+    ) ++ Seq(
```
Comment (wangyum) [inline project/SparkBuild.scala:233]: Will revert this change after Hive 2.3.7 is released.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -2423,7 +2423,8 @@ object SQLConf {
       .version("3.0.0")
       .stringConf
       .createWithDefault(
-        "https://maven-central.storage-download.googleapis.com/maven2/")
+        "https://maven-central.storage-download.googleapis.com/maven2/," +
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2426]: Will revert this change after Hive 2.3.7 is released.

**Comment by wangyum** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Nice @wangyum.

**Comment by HyukjinKwon** [comment]
@wangyum, could we test it against the actual RC - https://people.apache.org/~gates/apache-hive-2.3.7-rc0/ and https://repository.apache.org/content/repositories/orgapachehive-1100/?

**Comment by HyukjinKwon** [comment]
>How was this patch tested?
> Exist test.

Ideally we should test it out with the steps you mentioned before:

Hive side:

```
export JAVA_HOME=/usr/lib/jdk1.8.0_221
export PATH=$JAVA_HOME/bin:$PATH
cd /usr/lib/hive-2.3.5
bin/schematool -dbType derby -initSchema --verbose
bin/hive --service metastore
```

Spark side:

```
export JAVA_HOME=/usr/lib/jdk-11.0.3
export PATH=$JAVA_HOME/bin:$PATH
build/sbt clean package -Phive -Phadoop-3.2 -Phive-thriftserver
export SPARK_PREPEND_CLASSES=true
bin/spark-shell --conf spark.hadoop.hive.metastore.uris=thrift://localhost:9083
```

**Comment by dongjoon-hyun** [comment]
+1 for @HyukjinKwon 's comment. (https://github.com/apache/spark/pull/28148#issuecomment-610733977)

**Comment by wangyum** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by wangyum** [comment]
retest this please.

**Comment by dongjoon-hyun** [comment]
Finally! Thank you, @wangyum .

**Comment by wangyum** [comment]
retest this please.

**Comment by wangyum** [comment]
retest this please.

**Review by Fokko — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @wangyum and all.
I verified the basic syntax described at https://issues.apache.org/jira/browse/SPARK-29245 again on JDK11.
Merged to master/3.0.

**Comment by HyukjinKwon** [comment]
Nice, LGTM too.
